### ISSUE TITLE
Does the library support side-information for only a fraction of training examples?

### ISSUE THREAD

**Issue by asross** [tu-rbo/concarne#1]
This is half a question about the library and half a general side-information advice question -- apologies if I'm abusing the github issue feature. I have a problem in which I have side-information for a small but significant subset of my training data. If I train my model without this side-information, it doesn't generalize well, but the information needed to make it generalize well for the full dataset is contained in the side-info I have for my small subset.

I'm considering trying to use this library to make use of that information. It seems like it might be easiest if I frame the problem as multi-task learning, with 0 loss, 0 grad-loss for training examples with no corresponding side-information.

Does this seem like a sensical approach?

**Comment by shoefer** [comment]
Thanks for considering concarne for your problem! 

In principle, your idea of using side information only for a subset of your training totally makes sense. However, there is a simpler technical solution than the one you suggested. To address exactly this issue of unbalanced input/label and input/side-info data sets the PatternTrainer class features the fit_XZ_XY method. It should work with all the patterns currently implemented, so that you can easily play around with the different patterns and check which one works best for your problem. Depending on the pattern and the loss you use, you might have to play around a bit with the learning rates for the target and the side objective to get good performance.

For more information on how to use the PatternTrainer and fit_XZ_XY I recommend you to look at its [docstring](https://github.com/tu-rbo/concarne/blob/master/concarne/training.py) and check out the [handwritten.py](https://github.com/tu-rbo/concarne/blob/master/example/handwritten.py) example which uses exactly this feature.

Let me know if you have further questions!

**Comment by asross** [comment]
Awesome, thank you! That looks like it should work -- assuming I have enough side-information for even one minibatch :)

Thanks for creating this library, the idea is very interesting and I like `concarne`'s interface!

**Comment by shoefer** [comment]
You can always try minibatch size = 1 -- if that's still too small you're in trouble ;-)

To make a more serious comment: you can also use different mini batch sizes for XY and XZ by passing different values for `batch_size` and `XZ_batch_size` to the constructor of PatternTrainer.

You're welcome and good luck solving your problem!
